File: Editor/Tools/Scene/SaveSceneTool.cs
using System;
using UnityEngine;
using UnityEditor;
using UnityEditor.SceneManagement;
using UnityEngine.SceneManagement;
using McpUnity.Unity;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool to save the current scene
    /// Based on Unity API: EditorSceneManager.SaveScene
    /// https://docs.unity3d.com/ScriptReference/SceneManagement.EditorSceneManager.SaveScene.html
    /// </summary>
    public class SaveSceneTool : McpToolBase
    {
        public SaveSceneTool()
        {
            Name = "save_scene";
            Description = "Save the current scene or all open scenes.";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            try
            {
                bool saveAll = parameters["saveAll"]?.ToObject<bool>() ?? false;
                string savePath = parameters["savePath"]?.ToObject<string>();

                if (saveAll)
                {
                    // Save all open scenes
                    bool success = EditorSceneManager.SaveOpenScenes();
                    
                    JArray savedScenes = new JArray();
                    for (int i = 0; i < SceneManager.sceneCount; i++)
                    {
                        Scene scene = SceneManager.GetSceneAt(i);
                        savedScenes.Add(new JObject
                        {
                            ["name"] = scene.name,
                            ["path"] = scene.path
                        });
                    }

                    return new JObject
                    {
                        ["success"] = success,
                        ["message"] = success ? $"Saved {SceneManager.sceneCount} scene(s)." : "Failed to save some scenes.",
                        ["count"] = SceneManager.sceneCount,
                        ["savedScenes"] = savedScenes
                    };
                }
                else
                {
                    // Save active scene
                    Scene activeScene = SceneManager.GetActiveScene();
                    
                    bool success;
                    if (!string.IsNullOrEmpty(savePath))
                    {
                        // Save to new path
                        success = EditorSceneManager.SaveScene(activeScene, savePath);
                    }
                    else
                    {
                        // Save to current path
                        success = EditorSceneManager.SaveScene(activeScene);
                    }

                    return new JObject
                    {
                        ["success"] = success,
                        ["message"] = success ? $"Saved scene '{activeScene.name}'." : "Failed to save scene.",
                        ["sceneName"] = activeScene.name,
                        ["scenePath"] = activeScene.path
                    };
                }
            }
            catch (Exception ex)
            {
                return McpUnitySocketHandler.CreateErrorResponse($"Failed: {ex.Message}", "execution_error");
            }
        }
    }
}

File: Server~/src/tools/scene/SaveSceneTool.ts
import { z } from 'zod';
import { BaseTool } from '../base/BaseTool.js';
import { Tool, Tags } from '../base/ToolDecorators.js';

const SaveSceneToolArgsSchema = z.object({
  saveAll: z.boolean().optional().default(false).describe('Save all open scenes. Default: false (save active scene only)'),
  savePath: z.string().optional().describe('New path to save scene to (for Save As). Example: "Assets/Scenes/NewScene.unity"')
});

@Tool({
  name: 'save_scene',
  description: 'Save the current scene or all open scenes.',
  category: 'scene',
  version: '1.0.0'
})
@Tags(['unity', 'scene', 'save'])
export class SaveSceneTool extends BaseTool {
  get name() { return 'save_scene'; }
  get description() { return 'Save the current scene or all open scenes.'; }
  get inputSchema() { return SaveSceneToolArgsSchema; }
  get category() { return 'scene'; }
}
